subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/IndexPageGenerator.java

```diff
@@ -16,96 +16,238 @@
+        this.scenarios = readBuildResultData(resultJson);
+    }
+
+    private Set<ScenarioBuildResultData> readBuildResultData(File resultJson) {
+        try {
+            List<ScenarioBuildResultData> list = new ObjectMapper().readValue(resultJson, new TypeReference<List<ScenarioBuildResultData>>() {
+            });
+            return sortBuildResultData(list.stream().map(this::queryExecutionData));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @VisibleForTesting
+    Set<ScenarioBuildResultData> sortBuildResultData(Stream<ScenarioBuildResultData> data) {
+        Comparator<ScenarioBuildResultData> comparator = comparing(ScenarioBuildResultData::isBuildFailed).reversed()
+            .thenComparing(ScenarioBuildResultData::isSuccessful)
+            .thenComparing(comparing(ScenarioBuildResultData::isAboutToRegress).reversed())
+            .thenComparing(comparingDouble(ScenarioBuildResultData::getConfidencePercentage).reversed())
+            .thenComparing(comparingDouble(ScenarioBuildResultData::getRegressionPercentage).reversed())
+            .thenComparing(ScenarioBuildResultData::getScenarioName);
+        return data.collect(() -> new TreeSet<>(comparator), TreeSet::add, TreeSet::addAll);
+    }
+
+    private ScenarioBuildResultData queryExecutionData(ScenarioBuildResultData scenario) {
+        PerformanceTestHistory history = resultsStore.getTestResults(scenario.getScenarioName(), 3, 2, ResultsStoreHelper.determineChannel());
+        List<? extends PerformanceTestExecution> recentExecutions = history.getExecutions();
+        List<? extends PerformanceTestExecution> currentCommitExecutions = recentExecutions.stream().filter(execution -> execution.getVcsCommits().contains(commitId)).collect(Collectors.toList());
+        if (currentCommitExecutions.isEmpty()) {
+            scenario.setRecentExecutions(recentExecutions.stream().map(this::extractExecutionData).collect(Collectors.toList()));
+        } else {
+            scenario.setCurrentCommitExecutions(currentCommitExecutions.stream().map(this::extractExecutionData).collect(Collectors.toList()));
+        }
+
+        return scenario;
+    }
 
-    public IndexPageGenerator(List<NavigationItem> navigationItems) {
-        this.navigationItems = navigationItems;
+    private ScenarioBuildResultData.ExecutionData extractExecutionData(PerformanceTestExecution performanceTestExecution) {
+        List<MeasuredOperationList> nonEmptyExecutions = performanceTestExecution
+            .getScenarios()
+            .stream()
+            .filter(testExecution -> !testExecution.getTotalTime().isEmpty())
+            .collect(Collectors.toList());
+        if (nonEmptyExecutions.size() > 1) {
+            return new ScenarioBuildResultData.ExecutionData(performanceTestExecution.getStartTime(), getCommit(performanceTestExecution), nonEmptyExecutions.get(0), nonEmptyExecutions.get(1));
```

Comment:
What shall we show if we have more than one baseline? Does this code select the latest baseline (e.g. for the historical performance coordinator)?

Reply:
Currently the first non-empty version would be considered as baseline, the second non-empty version would be considered as current version. I was aware of multi-baseline, but didn't find any examples, could you please give me an example?

Reply:
See https://builds.gradle.org/repository/download/Gradle_Check_PerformanceExperimentCoordinator/15863660:id/report-performance-performance-tests.zip%21/report/tests/assemble-on-largeMonolithicJavaProject-with-local-cache-%28build-comparison%29.html

In the regular performance coordinator we don't have any example, only in the experiment coordinator and the historical performance. Maybe it is enough to consider the last entry as current and the pre-last entry as baseline?

See also https://builds.gradle.org/viewLog.html?buildId=15836212&buildTypeId=Gradle_Check_PerformanceHistoricalCoordinator&tab=report_project947_Performance

Reply:
Thanks. Let's use the last one as "current" and the pre-last one as "baseline" for now, and see if anyone complains.